File: back/src/crawling/crawling.service.ts
import { Injectable } from '@nestjs/common';
import axios from 'axios';
import * as cheerio from 'cheerio';
import { CreateMenuDto } from 'src/menus/dto/create-menu.dto';
import { Menu } from 'src/menus/entity/menu.entity';
import { MealType } from 'src/menus/enum/meal-type.enum';

@Injectable()
export class CrawlingService {
  async getPossibleDates() {
    const url = 'https://menu.payco.com/mrc/818490';
    const { data } = await axios.get(url);
    const $ = cheerio.load(data);

    const dates = [];
    $('.option.btnSelSvcYmd').each((_, element) => {
      const date = $(element).text().trim();
      dates.push(date);
    });

    return dates;
  }

  async getMenus(date: string) {
    const { data } = await axios.get(
      `https://menu.payco.com/service/shopMenu/menuList.nhn?shopMenuCfgSeq=29&serviceYmd=${date.replaceAll('-', '')}`,
    );
    const $ = cheerio.load(data);
    const menus: CreateMenuDto[] = [];

    $('.list_menu .item_menu').each((_, element) => {
      const name = $(element).find('.menu_title').text().trim();
      const descriptionHtml =
        $(element).find('.menu_desc').html()?.trim() || '';

      const description = descriptionHtml
        .replaceAll('\n', '')
        .replaceAll(/&amp;/g, '&')
        .split('<br>')
        .slice(0, -2)
        .join(', ');
      const caloriesText = $(element).find('.menu_cal').text().trim();
      const calories =
        parseInt(caloriesText.replace(/[^0-9]/g, ''), 10) || null; // 칼로리가 없으면 null

      const category = $(element).find('.menu_category').text().trim();
      const imageUrl = $(element).find('.menu_img_box img').attr('src') || null; // 이미지 URL이 없으면 null

      // MealType 결정
      let mealType: MealType;
      if (category.includes('중식')) {
        mealType = MealType.LUNCH;
      } else if (category.includes('석식')) {
        mealType = MealType.DINNER;
      } else {
        mealType = MealType.LUNCH; // 기본값
      }

      // 도시락 여부 확인
      const isLunchBox = category.includes('도시락');

      // Menu 객체 생성
      const menu: CreateMenuDto = {
        name,
        description: isLunchBox ? '' : description,
        calories,
        mealType,
        imageUrl,
        isLunchBox,
        date: new Date(date), // 날짜 고정 (변경 가능)
      };

      menus.push(menu);
    });

    console.log(
      'menus',
      menus.filter(
        (item, index, self) =>
          self.findIndex((t) => t.name === item.name) === index,
      ),
    );

    return menus;
  }
}

File: back-spring/src/main/java/com/nhn_eat/back/service/CrawlingService.java
package com.nhn_eat.back.service;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import com.nhn_eat.back.entity.MenuEntity;
import com.nhn_eat.back.entity.MealType;
import com.nhn_eat.back.repository.MenuRepository;

@Service
public class CrawlingService {

    @Autowired
    private MenuRepository menuRepository;

    public List<String> getPossibleDates() {
        String url = "https://menu.payco.com/mrc/818490";
        try {
            Document doc = Jsoup.connect(url).get();
            Elements dateElements = doc.select(".option.btnSelSvcYmd");
            List<String> dates = new ArrayList<>();
            dateElements.forEach(element -> dates.add(element.text().trim()));

            System.out.println("가능한 날짜 목록: " + dates);
            return dates;
        } catch (IOException e) {
            System.err.println("크롤링 중 오류 발생: " + e.getMessage());
        }

        return null;
    }


    public List<MenuEntity> getMenus(String date) {
        String url = "https://menu.payco.com/service/shopMenu/menuList.nhn?shopMenuCfgSeq=29&serviceYmd=" + date.replaceAll("-", "");
        List<MenuEntity> menus = new ArrayList<>();
        
        try {
            Document doc = Jsoup.connect(url).get();
            Elements menuElements = doc.select(".list_menu .item_menu");
            
            for (Element element : menuElements) {
                // 메뉴 이름 추출
                String name = element.select(".menu_title").text().trim();
                
                // 메뉴 설명 추출 및 처리
                String descriptionHtml = element.select(".menu_desc").html().trim();
                String description = processDescription(descriptionHtml);
                
                // 칼로리 추출
                String caloriesText = element.select(".menu_cal").text().trim();
                int calories = extractCalories(caloriesText);
                
                // 카테고리 추출
                String category = element.select(".menu_category").text().trim();
                
                // 이미지 URL 추출
                String imageUrl = element.select(".menu_img_box img").attr("src");
                if (imageUrl != null && imageUrl.startsWith("http:")) {
                    imageUrl = imageUrl.replace("http:", "https:");
                }
                
                // MealType 결정
                MealType mealType = determineMealType(category);
                
                // 도시락 여부 확인
                boolean isLunchBox = category.contains("도시락");
                
                // MenuEntity 생성
                MenuEntity menu = new MenuEntity();
                menu.setName(name);
                menu.setDescription(isLunchBox ? "" : description);
                menu.setCalories(calories);
                menu.setMealType(mealType);
                menu.setImageUrl(imageUrl);
                menu.setLunchBox(isLunchBox);
                menu.setDate(date);
                
                menus.add(menu);
            }
            
            System.out.println("크롤링된 메뉴 개수: " + menus.size());
            System.out.println(menus);
            return menus;
            
        } catch (IOException e) {
            System.err.println("크롤링 중 오류 발생: " + e.getMessage());
            return menus;
        }
    }
    
    private String processDescription(String descriptionHtml) {
        if (descriptionHtml == null || descriptionHtml.isEmpty()) {
            return "";
        }
        
        // HTML을 텍스트로 변환하고 처리
        String description = descriptionHtml
            .replaceAll("\n", "")
            .replaceAll("&amp;", "&")
            .replaceAll("<br>", ", ");
            
        // HTML 태그 제거
        description = description.replaceAll("<[^>]*>", "");
        
        // 마지막 쉼표들 제거
        if (description.endsWith(", ")) {
            description = description.substring(0, description.length() - 2);
        }
        
        return description.trim();
    }
    
    private int extractCalories(String caloriesText) {
        if (caloriesText == null || caloriesText.isEmpty()) {
            return 0;
        }
        
        // 숫자만 추출
        String numbers = caloriesText.replaceAll("[^0-9]", "");
        if (numbers.isEmpty()) {
            return 0;
        }
        
        try {
            return Integer.parseInt(numbers);
        } catch (NumberFormatException e) {
            return 0;
        }
    }
    
    private MealType determineMealType(String category) {
        if (category.contains("중식")) {
            return MealType.LUNCH;
        } else if (category.contains("석식")) {
            return MealType.DINNER;
        } else {
            return MealType.LUNCH; // 기본값
        }
    }

    // @Scheduled(fixedRate = 5000)
    public void syncMenu() {
        List<String> dates = this.getPossibleDates();
        for (String date : dates) {
            List<MenuEntity> crawlingMenus = this.getMenus(date).stream()
                .distinct()
                .collect(Collectors.toList());
            List<MenuEntity> previousMenus = menuRepository.findByDate(date);
            List<MenuEntity> noNeedMenus = previousMenus.stream()
                .filter(menu -> !crawlingMenus.contains(menu))
                .collect(Collectors.toList());
            List<MenuEntity> needUpdateMenus = previousMenus.stream()
                .filter(menu -> crawlingMenus.contains(menu))
                .collect(Collectors.toList());
            List<MenuEntity> newMenus = crawlingMenus.stream()
                .filter(menu -> !previousMenus.contains(menu))
                .collect(Collectors.toList());
            
            for (MenuEntity menu : noNeedMenus) {
                menuRepository.delete(menu);
            }
            for (MenuEntity newMenu : needUpdateMenus) {
                List<MenuEntity> existingMenus = menuRepository.findByNameAndDate(
                    newMenu.getName(), newMenu.getDate());

                if (existingMenus.size() > 0) {
                    MenuEntity menu = existingMenus.get(0);
                    menu.setDescription(newMenu.getDescription());
                    menu.setCalories(newMenu.getCalories());
                    menu.setImageUrl(newMenu.getImageUrl());
                    menu.setLunchBox(newMenu.isLunchBox());
                    menu.setLikeCount(0);
                    menuRepository.save(menu);
                } else {
                    menuRepository.save(newMenu);
                }
            }
            for (MenuEntity menu : newMenus) {
                menuRepository.save(menu);
            }
        }

    }
} 